### ISSUE TITLE
Env variables: how to?

### ISSUE THREAD

**Issue by ndrean** [dwyl/learn-elixir#211]
Desperate to understand how to use them.  It works locally but I am absolutely unable to deploy. I simply want to know where to put what.

You can put  "hard-coded" value in "config.exs":
```elixir
config :my_app,
    google_callback: "/gcb"
```
then `Application.get_env(:myapp, :google_callback)`  works (deployed)

Then when you call a `System.get_env("GOOGLE")` in the runtime? directly? Nothing works.
If I use `System.get_env("GOOGLE") ` somewhere in the code, it is not read.
Should you put it in "runtime.exs" or "prod.exs"? I looked at some examples but still don't understand. Any clean repo to understand where to put what? Many thanks.

**Comment by nelsonic** [comment]
@ndrean great question as ever. 

Anything you don't mind "leaking" i.e. being read by **`OpenAI`** (which you better believe has access to _all_ `private` GitHub repos because of their `MSFT` deal...) can be in `/config/prod.exs`. The sensitive things like API/AWS keys should be an environment variable.
A _second_ rule of thumb is: a variable that you don't want to require a code update for should be an environment variable too. e.g. a feature flag that you want to toggle in the environment and then just restart the app without re-deploying.
e.g. the `Debug Level` if you need to debug something _quickly_ without re-deploying an App you can simply toggle the `DEBUG="info"` environment variable, power-cycle the App and then _all_ the debugging will be enabled. Then once you're done debugging set it back to `DEBUG="error"` and cycle the app again. 

At least this is how we've done it in the past with bigger teams in companies with well-defined change-control processes.
i.e. All PRs even "debugging" ones have to be approved by 2 people and QA-tested before they can go to `production` so having the ability to toggle debugging as env var is _super_ useful.

**Comment by ndrean** [comment]
Nice, thks!  Yes, I pass most as env variables.

Nice tip for DEBUG

Second step is passing them in github secrets, later.

But my problem seemed even more simple. I can't pass the env vars to fly.io. looked in 2 repo:

- into <https://github.com/fly-apps/live_beats/blob/master/config/runtime.exs>.

They put the github credentials in "runtime.exs", populate with `System.fetch_env!` and calls with `Application.fetch_env!`:

```elixir
# runtime.exs
config :live_beats, :github,
    client_id: System.fetch_env!("LIVE_BEATS_GITHUB_CLIENT_ID"),
    client_secret: System.fetch_env!("LIVE_BEATS_GITHUB_CLIENT_SECRET")
```

When they want to use them, [they do](https://github.com/fly-apps/live_beats/blob/712204ca172ed6d36456c8ed4fd6aaba50099447/lib/live_beats/github.ex#L89):
```elixir
# github.ex
defp client_id, do: LiveBeats.config([:github, :client_id])
defp secret, do: LiveBeats.config([:github, :client_secret])
```

where `LiveBeats.config` is [basically does](https://github.com/fly-apps/live_beats/blob/712204ca172ed6d36456c8ed4fd6aaba50099447/lib/live_beats.ex#L20) for example:

```elixir
Application.fetch_env!(:live_beats, :github) |>  Keyword.fetch(:client_id)
```

- I looked into <https://github.com/dwyl/imgup/blob/main/config/runtime.exs>, same thing: 

AWS config is in "runtime.exs" but with a `System.get_env` this time, and the Env var [is called](https://github.com/dwyl/imgup/blob/main/lib/app/upload.ex) via a `Application.get_env`.


So, simple? Humm... here comes the famous  _this does not work for me_  🙄 yes yes!!. I don't know why, of course.

**Comment by nelsonic** [comment]
Yes, indeed https://github.com/dwyl/imgup/blob/a7d18ce6a4f0d3ceb512a5cdc494b2d7b3683050/config/runtime.exs#L67-L73 is a good example of how we use environment variables for `AWS` keys which we _definitely_ don't want to leak anywhere. 

What part is not working for you? 💭 
Are you sure the environment variables are available?
We have a simple checker in our MVP to confirm the variables are available: https://mvp.fly.dev/init
Ref: https://github.com/dwyl/mvp/blob/main/lib/app_web/controllers/init_controller.ex

**Comment by ndrean** [comment]
Yes, I did something similar, not that nice, but just print them on the landing page. Nada.

**Comment by ndrean** [comment]
Ah yes, I have another test: I need an env var to set up a module. If I simply use `System.get_env`, fly.io won't even compile. I only manage to compile if I hard code the env var.... means nothing passes, although they are set in fly.io.

**Comment by ndrean** [comment]
I can put this in "prod.exs" or "runtime.exs", it is not loaded:
```elixir
config :ex_aws,
  access_key_id: System.get_env("AWS_ACCESS_KEY_ID"),
  secret_access_key: System.get_env("AWS_SECRET_ACCESS_KEY"),
  region: System.get_env("AWS_REGION"),
  bucket: System.get_env("AWS_S3_BUCKET"),
  request_config_override: %{}

config :up_img, :github,
  github_client_id: System.get_env("GITHUB_CLIENT_ID"),
  github_client_secret: System.get_env("GITHUB_CLIENT_SECRET")

config :up_img, :google,
  google_client_id: System.get_env("GOOGLE_CLIENT_ID"),
  google_client_secret: System.get_env("GOOGLE_CLIENT_SECRET")

config :up_img, :vault_key, System.get_env("CLOAK_KEY")
```

The "reader" module does nothing more (or less) than previously shown:

```elixir
# reader.ex
def fetch_key(main, key),
    do:
      Application.fetch_env!(:up_img, main)  |> Keyword.get(key)

  def gh_id, do: fetch_key(:github, :github_client_id)
  def gh_secret, do: fetch_key(:github, :github_client_secret)

  def google_id, do: fetch_key(:google, :google_client_id)
  def google_secret, do: fetch_key(:google, :google_client_secret)

  def vault_key, do: Application.get_env(:up_img, :vault_key)

  def bucket, do: Application.get_env(:ex_aws, :bucket)
```

**Comment by nelsonic** [comment]
This is *very* strange. And feels like a support topic for the Fly/Elixir forum. 💭

**Comment by nelsonic** [comment]
Guaranteed your code is not "rubbish". 😜 
Did you use `fly launch` when setting up your app? 💭 
Without access to your `fly.toml` or `Dockerfile` can't see if it's loading the `env` correctly. 🤷‍♂️ 
My advice would be to re-run `fly launch` and re-create the deployment files.

**Comment by ndrean** [comment]
Well, I decided to run a DockerCompose since this is what fly does, and good news (or bad depends 😜), it breaks too! digging...

<img width="1440" alt="Screenshot 2023-09-19 at 16 31 51" src="https://github.com/dwyl/learn-elixir/assets/6793008/0347320b-9edd-4bd3-b87f-0f5b048757dc">

**Comment by ndrean** [comment]
I ran it under MIX_ENV=dev and moved the confi from "dev.exs" to "runtime.exs", and bad (?) news, it works on docker: the env vars are loaded and read.

<img width="932" alt="Screenshot 2023-09-19 at 16 41 21" src="https://github.com/dwyl/learn-elixir/assets/6793008/1a690de7-9f34-46c0-bd3e-93449fcc9ccd">

**Comment by ndrean** [comment]
I didn't change a single line to the code, config in "runtime.exs", as already shown .... and now it is deployed 🥵, at least the env vars are loaded.✌️. Didn't understand: 6 tries

<img width="560" alt="Screenshot 2023-09-19 at 19 10 04" src="https://github.com/dwyl/learn-elixir/assets/6793008/413eb44a-eb51-4c84-a7ca-8a75dc7dbbdb">

**Comment by ndrean** [comment]
Fun part: I used `:httpc` not to be depend of the HTTP client library used and thought it was part of Erlang, but:

```bash
Request: POST /google/callback
2023-09-19T17:34:22Z app[9185369a17d538] cdg [info]** (exit) an exception was raised:

2023-09-19T17:34:22Z app[9185369a17d538] cdg [info]    
** (UndefinedFunctionError) function :httpc.request/4 is undefined (module :httpc is not available)

2023-09-19T17:34:22Z app[9185369a17d538] cdg [info]        
:httpc.request(:get, {~c"https://www.googleapis.com/oauth2/v1/certs", []}, [], [])

2023-09-19T17:34:22Z app[9185369a17d538] cdg [info]        
(up_img 0.1.0) lib/libraries/google_certs.ex:67: ElixirGoogleCerts.fetch/1
```

**Comment by ndrean** [comment]
A good example (I think) of using "compiled" config is when you use a mock: with `Application.compile_env` as a module attribute. Declare it in "text.exs"  and "dev/prod.exs" and it will change upon the context. Nice in fact.

**Comment by ndrean** [comment]
You use this in ElixirGithubAuth.

https://github.com/dwyl/elixir-auth-github/blob/7e99a2e7574e0b86853575379d5765d3357b893d/lib/elixir_auth_github.ex#L12

**Comment by ndrean** [comment]
I was wondering if reading the env variable could be slow ?? Is it worth  loading them all in an ETS table when the app starts and retrieve them from ETS rather than reading? I implemented a Task to do so, it centralises the calls to get these env vars - could be from ETS or reading -  but it seems difficult to say if it's worth doing so. I imagine it is not worth and it was a trial to learn and run a Task :)
